src/chain/chain-generic-numerator.cc

```diff
@@ -221,35 +229,70 @@ bool GenericNumeratorComputation::ForwardBackward(
 
   bool ok = true;
-  Matrix<BaseFloat> alpha;
-  Matrix<BaseFloat> beta;
   Matrix<BaseFloat> probs;
-  Matrix<BaseFloat> derivs;
+  Matrix<BaseFloat> derivs; // Don't need nthreads copies to avoid data
+                            // races since each sequence operates on a
+                            // distinct set of columns
 
   // We selectively copy only those pdfs we need
   CopySpecificPdfsIndirect(nnet_output_, index_to_pdf_, &probs);
 
   derivs.Resize(probs.NumRows(), probs.NumCols());
   derivs.Set(-std::numeric_limits<BaseFloat>::infinity());
 
-  for (int seq = 0; seq < num_sequences; ++seq) {
-    // Forward part
-    AlphaFirstFrame(seq, &alpha);
-    partial_loglike += AlphaRemainingFrames(seq, probs, &alpha);
-
-    // Backward part
-    BetaLastFrame(seq, alpha, &beta);
-    BetaRemainingFrames(seq, probs, alpha, &beta, &derivs);
-    if (GetVerboseLevel() >= 1)
-      ok = ok && CheckValues(seq, probs, alpha, beta, derivs);
+  // Set total number of workers to the available hardware concurrency
+  unsigned int nthreads = opts_.multithreaded ? std::thread::hardware_concurrency() : 1;
+  // Naive load balancing, each thread gets a chunk of the sequences to process
+  unsigned int num_sequences_per_thread = 
+    (num_sequences + nthreads - 1) / nthreads;
+
+  // Allocate one alpha and beta matrix per thread to avoid contention
+  std::vector<Matrix<BaseFloat>> alpha(nthreads);
+  std::vector<Matrix<BaseFloat>> beta(nthreads);
+  
+  // Per thread partial values and boolean
+  std::vector<BaseFloat> partial_loglike_mt(nthreads, static_cast<BaseFloat>(0));
+  std::vector<bool> ok_mt(nthreads, true);
+
+  // Lambda function for each thread's portion of the computation
+  auto thread_lambda = [&] (int thread, int num_sequences, int num_sequences_per_thread) {
+    int seq_st = thread * num_sequences_per_thread;
+    int seq_en = seq_st + num_sequences_per_thread;
+    seq_en = (seq_en <= num_sequences) ? seq_en : num_sequences;
+    for (int seq = seq_st; seq < seq_en; ++seq) {
+      // Forward part
+      AlphaFirstFrame(seq, &alpha[thread]);
+      partial_loglike_mt[thread] += AlphaRemainingFrames(seq, probs, &alpha[thread]);
+
+      // Backward part
+      BetaLastFrame(seq, alpha[thread], &beta[thread]);
+      BetaRemainingFrames(seq, probs, alpha[thread], &beta[thread], &derivs);
+      if (GetVerboseLevel() >= 1)
+        ok_mt[thread] = ok_mt[thread] && CheckValues(seq, probs, alpha[thread], beta[thread], derivs);
+    }
+    return;
+  };
+
+  std::vector<std::thread> workers(nthreads);
+  for (int thread = 0; thread < nthreads; ++thread)
+    // Launch all threads
+    workers[thread] = std::thread(thread_lambda, thread, num_sequences, num_sequences_per_thread);
+  for (int thread = 0; thread < nthreads; ++thread) {
+    // Join threads back in
+    workers[thread].join();
+    // Reduce thread values to a single value
+    partial_loglike += partial_loglike_mt[thread];
+    ok = ok && ok_mt[thread];
```

Comment:
Does this actually create threads here, or uses some kind of thread pool?  Seems like creating threads might be kind of heavyweight.

Reply:
Yes, threads are created within `ForwardBackward(...)` and the number of threads created is set to be the available hardware concurrency and the total number of sequences are split across these threads. There is an overhead to creating threads this way, but since the `GenericNumeratorComputation` object is created once per iteration and `ForwardBackward` is called only once during its lifetime, this design seemed reasonable.

Reply:
I'm a bit worried about setting these things automatically -- hardware concurency is one thing and user limits placed on resources (e.g. in SGE/Slurm) are other things. This reminds me the OMP mentality that causes issues in many similar environments.
So I think it would be a nice thing to have but I don't think it should be default.

Reply:
So, make this also opt-in? Or set a better default value?